protoc_plugin/lib/src/protobuf_field.dart

```diff
@@ -63,9 +64,32 @@ class ProtobufField {
   bool get isRepeated =>
       descriptor.label == FieldDescriptorProto_Label.LABEL_REPEATED;
 
-  /// True if the field is to be encoded with [packed=true] encoding.
-  bool get isPacked =>
-      isRepeated && descriptor.options != null && descriptor.options.packed;
+  /// True if a numeric field is repeated and it should be encoded with packed
+  /// encoding. In proto3 repeated fields are encoded as packed by default.
+  /// proto2 requires `[packed=true]` option.
+  bool get isPacked {
+    if (!isRepeated) {
+      return false; // only repeated fields can be packed
+    }
+
+    if (!baseType.isPackable) {
+      return false;
+    }
+
+    if (parent.fileGen.descriptor.syntax == 'proto3') {
```

Comment:
> Do we have more places we check for proto3 syntax?

This is the only place where we check the syntax as far as I can see. We have some commented-out pseudo-code here: https://github.com/google/protobuf.dart/blob/ad3f06f759087bbef1f7e78cf7b6185039ae74ff/protoc_plugin/lib/src/protobuf_field.dart#L126-L128

I wonder why the `syntax` field in `descriptor.proto` is a `string` and not an `enum`..

> Perhaps we should wrap it in a getter FileGen.syntax returning an enum or something like that?

Yeah I think this makes sense. I'll add it.

Reply:
> Yeah I think this makes sense. I'll add it.

The code is not nice without null safety:

```dart
  bool get isPacked {
    if (!isRepeated) {
      return false; // only repeated fields can be packed
    }

    if (!baseType.isPackable) {
      return false;
    }

    switch (parent.fileGen.protoSyntax) {
      case ProtoSyntax.proto3:
        if (!descriptor.hasOptions()) {
          return true; // packed by default in proto3
        } else {
          return !descriptor.options.hasPacked() || descriptor.options.packed;
        }
      case ProtoSyntax.proto2:
        if (!descriptor.hasOptions()) {
          return false; // not packed by default in proto3
        } else {
          return descriptor.options.packed;
        }
    }
  }
```

I have to add a `break` after the first `case`, and a `return` after the `switch`.

Should we enable null safety in protoc_plugin?

Reply:
> Should we enable null safety in protoc_plugin?

This was done in #642 so I introduced an enum and the code is nice now.

However I also realized a potential problem. Some of the `ProtobufField` methods (`generateDefaultFunction`, `generateBuilderInfoCall`, `getDartType`) take a `FileGenerator` argument instead of using `parent.fileGen`. It seems like the assumption here is that a single `ProtobufField` instance can be used with different `FileGenerator`s. I think this assumption is probably wrong and we should be able to eliminate those `FileGenerator` args and always use `parent.fileGen`. However if the assumption is right then our use of `parent.fileGen` in this PR is incorrect.

I will check if a `ProtobufField` is ever used with multiple `FileGenerator`s. I will be surprised if it is.

Reply:
Ah, that's because we didn't have the `parent` field before. I will update the rest of the code to use `parent`.